File: database-driver-plugin/src/main/java/de/lino/database/utility/cache/DefaultConsistentHashRing.java
package de.lino.database.utility.cache;

import de.lino.database.utils.cache.ConsistentHashRing;
import lombok.NonNull;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.NavigableMap;
import java.util.concurrent.ConcurrentSkipListMap;

/**
 * Default implementation of {@link ConsistentHashRing}, implemented in pure
 * Java (only {@link java.security.MessageDigest} for SHA-256, part of the JDK).
 * <p>
 * Each physical node is assigned several "virtual nodes" on the ring, which
 * spreads load more evenly and avoids hotspots. A key is hashed onto the ring
 * and routed to the next node clockwise.
 * <p>
 * <b>Complexity:</b> {@link #nodesFor} is O(log V) for the ring lookup
 * (V = total number of virtual nodes) plus O(R) for replica selection —
 * independent of the number of stored keys n. That matters for scaling: whether
 * there are 10<sup>3</sup> or 10<sup>12</sup> keys, routing itself stays equally fast.
 * <p>
 * <b>Thread-safety:</b> the ring is backed by a {@link ConcurrentSkipListMap} so
 * that {@link #addNode} / {@link #removeNode} can safely run concurrently with
 * lookups (e.g. while rebalancing a live cluster), and the per-thread cached
 * {@link MessageDigest} in {@link #hash} avoids repeating the relatively costly
 * algorithm lookup on every call on this hot path.
 *
 * @param <NodeId> type used to identify a node
 */
public final class DefaultConsistentHashRing<NodeId> implements ConsistentHashRing<NodeId> {

    private static final int VIRTUAL_NODES_PER_NODE = 100;

    private static final ThreadLocal<MessageDigest> DIGEST = ThreadLocal.withInitial(() -> {
        try {
            return MessageDigest.getInstance("SHA-256");
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 not available", e); // cannot happen on a standard JVM
        }
    });

    private final NavigableMap<Long, NodeId> ring = new ConcurrentSkipListMap<>();

    /**
     * @param nodes initial set of nodes to place on the ring, must not be {@code null} or empty
     * @throws IllegalArgumentException if {@code nodes} is empty
     */
    public DefaultConsistentHashRing(@NonNull List<NodeId> nodes) {
        if (nodes.isEmpty()) {
            throw new IllegalArgumentException("at least one node is required");
        }
        for (NodeId node : nodes) {
            addNode(node);
        }
    }

    @Override
    public void addNode(@NonNull NodeId node) {
        for (int i = 0; i < VIRTUAL_NODES_PER_NODE; i++) {
            ring.put(hash(node.toString() + "#vnode" + i), node);
        }
    }

    @Override
    public void removeNode(@NonNull NodeId node) {
        ring.entrySet().removeIf(e -> e.getValue().equals(node));
    }

    @Override
    public NodeId nodeFor(@NonNull Object key) {
        long h = hash(String.valueOf(key));
        Map.Entry<Long, NodeId> entry = ring.ceilingEntry(h);
        if (entry == null) {
            entry = ring.firstEntry(); // wrap around the ring
        }
        return entry.getValue();
    }

    @Override
    public List<NodeId> nodesFor(@NonNull Object key, int replicationFactor) {
        long h = hash(String.valueOf(key));
        NavigableMap<Long, NodeId> tail = ring.tailMap(h, true);

        List<NodeId> result = new ArrayList<>(replicationFactor);
        for (NodeId candidate : tail.values()) {
            addIfAbsent(result, candidate, replicationFactor);
            if (result.size() == replicationFactor) return result;
        }
        for (NodeId candidate : ring.values()) { // wrap around once the end of the ring is reached
            addIfAbsent(result, candidate, replicationFactor);
            if (result.size() == replicationFactor) break;
        }
        return result;
    }

    private void addIfAbsent(List<NodeId> result, NodeId candidate, int max) {
        if (!result.contains(candidate) && result.size() < max) {
            result.add(candidate);
        }
    }

    private long hash(String input) {
        byte[] bytes = DIGEST.get().digest(input.getBytes(StandardCharsets.UTF_8));
        // interpret the first 8 bytes of the digest as a long -> uniformly distributed hash
        long result = 0;
        for (int i = 0; i < 8; i++) {
            result = (result << 8) | (bytes[i] & 0xff);
        }
        return result;
    }
}

File: python/database-driver-plugin/src/database_driver/plugin/utility/cache/default_consistent_hash_ring.py
"""Mirror of ``de.lino.database.utility.cache.DefaultConsistentHashRing``."""

from __future__ import annotations

import bisect
import hashlib
import threading
from collections.abc import Sequence
from typing import Any, TypeVar

from database_driver.api.utils.cache.consistent_hash_ring import ConsistentHashRing

NodeId = TypeVar("NodeId")

_VIRTUAL_NODES_PER_NODE = 100


class DefaultConsistentHashRing(ConsistentHashRing[NodeId]):
    """Default implementation of ``ConsistentHashRing``, in pure Python (only
    :mod:`hashlib` for SHA-256, part of the stdlib).

    Each physical node is assigned several "virtual nodes" on the ring, which spreads
    load more evenly and avoids hotspots. A key is hashed onto the ring and routed to the
    next node clockwise.

    Complexity: :meth:`nodes_for` is O(log V) for the ring lookup (V = total number of
    virtual nodes) plus O(R) for replica selection - independent of the number of stored
    keys n. Whether there are 10^3 or 10^12 keys, routing itself stays equally fast.

    Thread-safety: the ring is a sorted list guarded by one lock - the Python stand-in
    for the Java edition's ``ConcurrentSkipListMap`` - so :meth:`add_node` /
    :meth:`remove_node` can safely run concurrently with lookups (e.g. while rebalancing
    a live cluster). SHA-256 needs no per-thread digest caching here: :func:`hashlib.sha256`
    construction is cheap in CPython, unlike the JVM's ``MessageDigest`` lookup.
    """

    def __init__(self, nodes: Sequence[NodeId]) -> None:
        """Args:
            nodes: Initial set of nodes to place on the ring; must not be empty.

        Raises:
            ValueError: If ``nodes`` is empty.
        """
        if not nodes:
            raise ValueError("at least one node is required")
        self._lock = threading.Lock()
        self._hashes: list[int] = []
        self._nodes: list[NodeId] = []
        for node in nodes:
            self.add_node(node)

    def add_node(self, node: NodeId) -> None:
        if node is None:
            raise TypeError("node must not be None")
        with self._lock:
            for i in range(_VIRTUAL_NODES_PER_NODE):
                position = _hash(f"{node}#vnode{i}")
                index = bisect.bisect_left(self._hashes, position)
                # A hash collision overwrites the colliding slot, matching the Java
                # NavigableMap's put-on-same-key semantics.
                if index < len(self._hashes) and self._hashes[index] == position:
                    self._nodes[index] = node
                else:
                    self._hashes.insert(index, position)
                    self._nodes.insert(index, node)

    def remove_node(self, node: NodeId) -> None:
        if node is None:
            raise TypeError("node must not be None")
        with self._lock:
            keep = [(position, owner) for position, owner in zip(self._hashes, self._nodes) if owner != node]
            self._hashes = [position for position, _ in keep]
            self._nodes = [owner for _, owner in keep]

    def node_for(self, key: Any) -> NodeId:
        if key is None:
            raise TypeError("key must not be None")
        position = _hash(str(key))
        with self._lock:
            index = bisect.bisect_left(self._hashes, position)
            if index == len(self._hashes):
                index = 0  # wrap around the ring
            return self._nodes[index]

    def nodes_for(self, key: Any, replication_factor: int) -> list[NodeId]:
        if key is None:
            raise TypeError("key must not be None")
        position = _hash(str(key))
        result: list[NodeId] = []
        with self._lock:
            start = bisect.bisect_left(self._hashes, position)
            total = len(self._nodes)
            for step in range(total):  # walk clockwise, wrapping once around
                candidate = self._nodes[(start + step) % total]
                if candidate not in result and len(result) < replication_factor:
                    result.append(candidate)
                if len(result) == replication_factor:
                    break
        return result


def _hash(value: str) -> int:
    """Interprets the first 8 bytes of the value's SHA-256 digest as an unsigned integer
    - a uniformly distributed ring position, mirroring the Java edition's digest-to-long
    fold."""
    return int.from_bytes(hashlib.sha256(value.encode("utf-8")).digest()[:8], "big")
